flow/aggregates.py

```diff
@@ -0,0 +1,475 @@
+            when greater flexibility is needed.
+        :type key:
+            str, Iterable, or callable
+        :param default:
+            Default values used for grouping if invalid key is passed.
+        :type default:
+            str, Iterable, or callable
+        :param sort_by:
+            Before aggregating, sort the jobs by a given statepoint parameter.
+            The default behavior is no sorting.
+        :type sort_by:
+            str or NoneType
+        :param reverse_order:
+            States if the jobs are to be sorted in reverse order.
+            The default value is False.
+        :type reverse_order:
+            bool
+        :param select:
+            Condition for filtering individual jobs. This is passed as the
+            callable argument to `filter`.
+            The default behavior is no filtering.
+        :type select:
+            callable or NoneType
+        """
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+        elif isinstance(key, Iterable):
+            keys = list(key)
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                if isinstance(default, Iterable):
+                    if len(default) != len(keys):
+                        raise ValueError("Expected length of default argument is "
+                                         f"{len(keys)}, got {len(default)}.")
+                else:
+                    raise TypeError("Invalid default argument. Expected Iterable, "
+                                    f"got {type(default)}")
+
+                def keyfunction(job):
+                    return [job.sp.get(key, default_value)
+                            for key, default_value in zip(keys, default)]
+        elif callable(key):
+            keyfunction = key
+        else:
+            raise TypeError("Invalid key argument. Expected either str, Iterable "
+                            f"or a callable, got {type(key)}")
+
+        def aggregator_function(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator_function, sort_by, reverse_order, select, f'groupby{key}-{default}')
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+        elif (
+             self._sort_by != other._sort_by or
+             self._reverse_order != other._reverse_order or
+             self._tag != other._tag
+        ):
+            return False
+        elif (
+             self._select == other._select and
+             self._aggregator_function == other._aggregator_function
```

Comment:
I can't seem to access the original thread of discussion here, so I'll just make a new comment. The point that I was trying to make about this function (the whole `__eq__` method) being complicated was not about why we need the `tag`. I get that the tag might be necessary to distinguish between two. I was asking whether this `elif` clause is worth having at all. The remainder of this method basically tries to compare either the `co_code` or the hash of the `select`s and `aggregator_function`s to determine their equality, and either of those will be basically the same speed as the equality check in this `elif` (the `co_code` is just cached in the object on creation AFAIK). So I don't see any reason not to change this `elif` to an `else` clause, remove the equality checks that are currently here, and move the code with `self_select`, `other_select`, etc into this `else` clause.

Reply:
@vyasr, @b-butler asked me to check this first in order to reduce the complexity.
For example, if we're able to compare the `aggregator_function` and `select` directly then there's no need to compare the binary code. But if we fail, then we end up comparing the binary code.

Reply:
I understand, but I'm saying that I don't expect this to lead to any real performance improvement. In practice I expect users to use the provided `groupsof` and `groupby` options the vast majority of the time, and those will never pass this check because those decorators create new internal functions each time they are called. When users do bother to create custom aggregation functions, I expect them to use lambdas or something most of the time, in which case again this equality check will fail. While checking the type and then checking the sort/reverse/tag are probably useful, this branch adds complexity without any evidence that it will improve performance. That is the hallmark of premature optimization, so I'm recommending not doing this unless we see that the `__eq__` method is actually a performance bottleneck that could benefit from this extra check.

Reply:
I understand what you say.
@b-butler It'd be really helpful to know your views on this. If you agree, then I'll go ahead and remove the block.

Reply:
I agree with @vyasr here generally, hence the comment I made before.